File: python/year_2015/day_23.py
"""
https://adventofcode.com/2015/day/23
"""


from pathlib import Path
from functools import partial
from itertools import takewhile
from operator import gt

from iteration_utilities import applyfunc, consume

def compute(text, a, b, target):
    def process(line):
        instruction, *operands = line.replace(",", "").split()
        if instruction == "jmp":
            return instruction, None, int(operands[0]) - 1
        elif len(operands) == 1:
            return instruction, operands[0], None
        else:
            return instruction, operands[0], int(operands[1]) - 1

    def execute(pointer):
        instruction, register, offset = program[pointer]

        if instruction == "hlf":
            registers[register] //= 2
        elif instruction == "tpl":
            registers[register] *= 3
        elif instruction == "inc":
            registers[register] += 1
        elif instruction == "jmp":
            pointer += offset
        elif instruction == "jie" and registers[register] % 2 == 0:
            pointer += offset
        elif instruction == "jio" and registers[register] == 1:
            pointer += offset

        return pointer + 1

    registers = {"a": a, "b": b}
    program = tuple(map(process, text.splitlines()))

    consume(takewhile(partial(gt, len(program)), applyfunc(execute, 0)), None)

    return registers[target]


def part_1(text, a=0, b=0, target="b"):
    return compute(text, a, b, target)


def part_2(text, a=1, b=0, target="b"):
    return compute(text, a, b, target)


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_23.txt").read_text()


EXAMPLE = """inc a
jio a, +2
tpl a
inc a
"""


def test_part_1_input():
    assert part_1(_read_input()) == 307

def test_part_1_example_a():
    assert part_1(EXAMPLE, target='a') == 2

def test_part_1_example_b():
    assert part_1(EXAMPLE, target='b') == 0

def test_part_2_input():
    assert part_2(_read_input()) == 160

def test_part_2_example_a():
    assert part_2(EXAMPLE, a=2, target='a') == 10

def test_part_2_example_b():
    assert part_2(EXAMPLE, target='b') == 0

File: rust/year_2015/src/day_23.rs
use Instruction::*;
use num::{Integer as _, One as _};

pub fn part_1(input: &str) -> u32 {
    run(input, 0)
}

pub fn part_2(input: &str) -> u32 {
    run(input, 1)
}

fn run(input: &str, reg_a: u32) -> u32 {
    let instructions = parse_instructions(input);
    let mut registers = [reg_a, 0];
    let mut pc: usize = 0;
    while let Some(instruction) = instructions.get(pc) {
        let jump = match *instruction {
            Hlf(r) => {
                registers[r] /= 2;
                1
            }
            Tpl(r) => {
                registers[r] *= 3;
                1
            }
            Inc(r) => {
                registers[r] += 1;
                1
            }
            Jmp(offset) => offset,
            Jie(r, offset) => {
                if registers[r].is_even() {
                    offset
                } else {
                    1
                }
            }
            Jio(r, offset) => {
                if registers[r].is_one() {
                    offset
                } else {
                    1
                }
            }
        };
        pc = pc.strict_add_signed(jump)
    }
    registers[B]
}

type Register = usize;

const A: Register = 0;
const B: Register = 1;

#[derive(Copy, Clone)]
enum Instruction {
    Hlf(Register),
    Tpl(Register),
    Inc(Register),
    Jmp(isize),
    Jie(Register, isize),
    Jio(Register, isize),
}

fn parse_instructions(input: &str) -> Vec<Instruction> {
    fn reg(reg: &str) -> Register {
        match reg {
            "a" => A,
            "b" => B,
            _ => panic!("invalid register: {reg}"),
        }
    }
    input
        .lines()
        .map(|line| {
            let (op, rest) = line.split_once(' ').unwrap();
            match op {
                "hlf" => Hlf(reg(rest)),
                "tpl" => Tpl(reg(rest)),
                "inc" => Inc(reg(rest)),
                "jmp" => Jmp(rest.parse().unwrap()),
                "jie" => {
                    let (r, offset) = rest.split_once(", ").unwrap();
                    Jie(reg(r), offset.parse().unwrap())
                }
                "jio" => {
                    let (r, offset) = rest.split_once(", ").unwrap();
                    Jio(reg(r), offset.parse().unwrap())
                }
                _ => panic!("invalid instruction: {line}"),
            }
        })
        .collect()
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_23.txt");

    #[test_case(INPUT => 307)]
    fn part_1(input: &str) -> u32 {
        super::part_1(input)
    }

    #[test_case(INPUT => 160)]
    fn part_2(input: &str) -> u32 {
        super::part_2(input)
    }
}
